New ui infinite scroll issues B087, B088, B089, B090

juzen2003 opened the pull request:
# Update
- B087: After resizing window and toggling between table & gallery, slider value will not change now.
- B088: InfiniteScroll scroll up:
    - The previous scroll up issue:
        1. When we scroll the slider to the middle, and hold down pageUp key to scroll up.
        2. When scrollbar reaches to the up scroll threshold point, in "ps-scroll-up" event handler function, infiniteScroll loadNextPage will be triggered with the standard steps:
            a. infiniteScroll request
            b. infiniteScroll load
            c. infiniteScroll load event handler (renderGalleryAndTable is called here)
        3. And after that it will only have one loadNextPage triggered (in step 2), and scrollbar will be stuck at the top end. No more data will be prefetched unless we use the wheel or lower the scroll bar a bit and move it up again.
    - The Reason:
        In step 2, we are still holding down pageUp key, and still getting into "ps-scroll-up" event to call the handler function to perform loadNextPage for a second time. However, when we try to call loadNextPage for the 2nd time, the first trigger of loadNextPage is still in the middle of 2-a and 2-b. Therefore, the 2nd time loadNextPage will not be called because infiniteScroll thinks it's already been doing loadNextPage (in the middle of 2-a and 2-b). And the scrollbar will be stuck at the top end.
    - Now in checkScroll, we will trigger "ps-scroll-up" when scrollbar reaches to up scroll threshold point.
- B089: We set the fixed PS scrollbar length to 100 for now.
- B090: 0.5S debounce time is removed.

juzen2003 committed:
Commit before switching branches

juzen2003 committed:
Merge remote-tracking branch 'upstream/new_ui' into new_ui_infiniteScroll_issues

juzen2003 committed:
Update code in renderGalleryAndTable, onUpdateSlider, and event handler function for ps-scroll-up

juzen2003 committed:
Add galleryAndTabPSLength to fix the ps length in gallery & table

juzen2003 committed:
Update checkScroll to trigger ps-scroll-up when it's above scroll up threshold point, this will alow us to hold down page up key and keep data loading

juzen2003 committed:
Clean up codes, commit before updating updateSliderHandle

juzen2003 committed:
updateSliderHandle with math for both gallery & table view to obtain startObs

juzen2003 committed:
Commit before replacing loop with math in updateSliderHandle

juzen2003 committed:
Update updateSliderHandle, replace loop with math

juzen2003 committed:
Update the math in updateSliderHandle, now after resizing the window, when switching between table and gallery, slider value will not change

juzen2003 committed:
Clean up codes

juzen2003 committed:
Rename galleryAndTabPSLength to galleryAndTablePSLength

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -14,6 +14,8 @@ const tableSortUpArrow = "fas fa-sort-up";
 const defaultTableSortArrow = "fas fa-sort";
 const infiniteScrollUpThreshold = 100;
+// Fixed scrollbar length for gallery & table view
+const galleryAndTabPSLength = 100;
```
Rename to galleryAndTablePSLength

dstopp replied:
a better variable name might be galleryAndTablePSLength

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -55,7 +59,8 @@ var o_browse = {
         // note: using .on vs .click allows elements to be added dynamically w/out bind/rebind of handler
 
-        $(".op-gallery-view, .op-dataTable-view").on("scroll", _.debounce(o_browse.checkScroll, 500));
+        // $(".op-gallery-view, .op-dataTable-view").on("scroll", _.debounce(o_browse.checkScroll, 500));
```
Remove old code

dstopp replied:
do we still need this?

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -635,12 +638,13 @@ var o_browse = {
         let elem = $(`#${opus.prefs.view} .thumbnail-container[data-obs="${value}"]`);
         let startObsLabel = o_browse.getStartObsLabel();
 
+        // Update obsNum in infiniteScroll instances, and this obsNum is the first item in current page. (will be used to set scrollbar position in renderGalleryAndTable).
```
Please break the long comments across multiple lines so they don't go past the edge of a reasonable browser width.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -635,12 +638,13 @@ var o_browse = {
-            $(`${tab} .op-dataTable-view`).infiniteScroll({"obsNum": value});
-            opus.prefs[startObsLabel] = value;
         } else {
             // When scrolling on slider and loadData is called, we will fetch 3 * getLimit items (one current page, one next page, and one previous page) starting from obsNum.
```
Same for this comment

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -666,47 +666,52 @@ var o_browse = {
-                let obsNum = $(elem).data("obs");
+            // table: obsNum = firstCachedObs + number of row
+            // gallery: obsNum = firstCachedObs + number of row * number of obs in a row
+            let obsNumDiff = (opus.prefs.browse === "dataTable") ? Math.round((topBoxBoundary - firstCachedObsTop)/$(`${tab} tbody tr`).outerHeight()) : Math.round((topBoxBoundary - firstCachedObsTop)/o_browse.imageSize) * o_browse.galleryBoundingRect.x;
```
Break into multiple lines so it's more readable

rfrenchseti replied:
Also, this won't work for the cart, will it? opus.prefs.browse is only for the main browse tab. Need to use Debby's new functions that retrieve the proper browse name.

juzen2003 committed:
Merge with upstream/new_ui

juzen2003 committed:
Clean up codes

juzen2003 committed:
Put back the correct math in updateSliderHandle

juzen2003 committed:
Update updateSliderHandle to delete some obs to meet the new row boundary when resizing

juzen2003 committed:
Clean up codes

juzen2003 committed:
Remove unwanted comments

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -675,56 +674,86 @@ var o_browse = {
+
+            // table: obsNum = calculatedFirstObs + number of row
+            // gallery: obsNum = calculatedFirstObs + number of row * number of obs in a row
+            let obsNumDiff = ((opus.prefs[browse] === "dataTable") ?
```
i don't think this is "dataTable" - I believe this is "data".

rfrenchseti replied:
Actually, no. The slug value in the URL is data, but the value in opus.prefs is dataTable. This is something Debby put in a while back when she didn't realize it would cause backwards compatibility problems with the URL. We fixed it in hash.js by doing the mapping when parsing/creating the URL. I hate this. And I just noticed a bug - hash.js only does the mapping for 'browse', not 'cart_browse'. See hash.js line 147. This needs to be fixed. I would MUCH rather get rid of dataTable entirely and just use data, though.

juzen2003 replied:
All dataTable related stuff is removed, now in slug, we will have data in browse or cart_browse when we are in table view.

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -675,56 +674,86 @@ var o_browse = {
+            // Update obsNum in both infiniteScroll instances.
+            // Store the most top left obsNum in gallery for both infiniteScroll instances
+            // (this will be used to updated slider obsNum).
+            if (contentsView === ".op-data-table-view") {
```
i find it confusing to check the container class vs. whether this is gallery or table view.

rfrenchseti replied:
Agreed

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -675,56 +674,86 @@ var o_browse = {
+            // delete first "numToDelete" obs if row size is changed
+            if (numToDelete !== 0) {
+                for (let count = 0; count < numToDelete; count++) {
+                    o_browse.deleteCachedObservation(galleryObsElem, tableObsElem, count);
```
I'm a little confused by this. When we delete obs 0, doesn't that make the next first obs also become obs 0? So we should be deleting obs 0 "numToDelete" times rather than deleting increasing numbers? I think this has the effect of deleting obs 0, 1, 3, 6, ... (original numbers before the delete). Actually the same comment applies to manageObservationCache. Am I confused or does this function do the wrong thing?

juzen2003 replied:
The current function will do the right thing. When we pass in the galleryObsElem and tableObsElem, they are arrays of each thumbnail container and table tr. When deleting an element in an array, the element will be removed and replaced with an undefined item like the following example:
```js
let a = [6, 7, 8, 9, 10];
delete a[0];
delete a[1];
console.log(a); // [ <2 empty items>, 8, 9, 10 ]
console.log(a.length); // 5
for (let i = 0; i < a.length; i++) {
    console.log(a[i]); 
}
// the above loop will print out the following:
// undefined
// undefined
// 8
// 9
// 10
``` 
The case you mentioned above (to always delete index 0) will work if we don't pass in galleryObsElem and tableObsElem, and always re-declare both variables inside the deleteCachedObservation function like this:
```js
deleteCachedObservation: function (galleryObsElem, tableObsElem, index) {
    let galleryObsElem = $(`${tab} .gallery [data-obs]`);
    let tableObsElem = $(`${tab} .op-data-table-view [data-obs]`);
// ...
```
This will create a new array without the undefined (previous deleted obs) items.

rfrenchseti replied:
OK, I understand what's going on now. It works fine. Thanks.

rfrenchseti reviewed (changes requested):
- (B087) I still see the slider changing when I switch between the gallery and table views with almost any size browser, even when I haven't done a resize since the last reload
- B088 appears fixed
- B089 appears fixed
- B090 appears fixed

juzen2003 committed:
Remove dataTable, and use data for browse view when we are in table view

juzen2003 committed:
Update to only execute resize logic in updateSliderHandler when it's triggered from window resize

juzen2003 committed:
Fix code style

juzen2003 committed:
Update adjustBrowseHeight

juzen2003 committed:
Update calculateGalleryHeight to fix the gallery and table view height in the cart

juzen2003 committed:
Clean up codes

juzen2003 committed:
Remove spinner in cart tab for infiniteScroll if all data are rendered

juzen2003 committed:
Clean up code style

juzen2003 committed:
Update adjustProductInfoHeight

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -660,71 +664,100 @@ var o_browse = {
 
     // find the first displayed observation index & id in the upper left corner
-    updateSliderHandle: function() {
+    updateSliderHandle: function(resize=false) {
```
"resize" makes this sound like something updateSlider should do itself. But this is called when a resize has already happened. How about something like "browserResized"?

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -660,71 +664,100 @@ var o_browse = {
+                // At this point of time, galleryBoundingRect is updated with new row size
+                // from countGalleryImages in adjustBrowseHeight.
+                let numToDelete = ((galleryBoundingRect.x - (firstCachedObs - 1) % galleryBoundingRect.x) %
+                galleryBoundingRect.x);
```
Indent to be past above open paren

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1475,7 +1506,8 @@ var o_browse = {
                 if ((firstObs === 1 && infiniteScrollData.options.loadPrevPage === true) ||
-                    (lastObs === opus.resultCount && infiniteScrollData.options.loadPrevPage === false)) {
+                    (lastObs === opus.resultCount && infiniteScrollData.options.loadPrevPage === false) ||
+                    (lastObs === o_cart.cartCount && infiniteScrollData.options.loadPrevPage === false)) {
```
This seems like it can't work. You're always comparing lastObs to BOTH the result count and the cart count, regardless of which tab you're actually dealing with. So conceivably you could be equal to the cart count while looking at the browse page, and that's going to mess things up, right?

juzen2003 replied:
lastObs is also depending on tab (the tab was defined in the callback function when we first init infiniteScroll). When it's in browse gallery view, it's trying to select the last rendered item in browse gallery. When it's in cart gallery view, it's trying to select the last rendered item in cart gallery.

rfrenchseti replied:
But you could be on the browse tab and lastObs could equal the cart count, which would be wrong. Or vice verse. Checking for resultcount or cartcount won’t work by itself. You need to check the tab name also. Also it would be good to pull the loadPrevPage out so you aren’t checking for false in each case. Would be easier to read.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1619,7 +1651,10 @@ var o_browse = {
         let footerHeight = $(".app-footer").outerHeight();
         let mainNavHeight = $("#op-main-nav").outerHeight();
-        let navbarHeight = $(".panel-heading").outerHeight();
+        // we need to specify the tab here
```
Do we need this comment? We select using tab all over the place without explaining why.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1637,9 +1672,9 @@ var o_browse = {
     },
 
-    adjustBrowseHeight: function() {
+    adjustBrowseHeight: function(resize=false) {
```
Same comment on "resize" naming as above

rfrenchseti reviewed (changes requested):
1) The Y scrollbar on the gallery is often missing from the browse tab using Chrome. I have no idea why.
2) The slider is misbehaving badly. When you drag it it doesn't stay on the value you select. The gallery and table don't go to the right obs. And when you switch back and forth between table and gallery the slider changes value.

juzen2003 committed:
- In browse.js, change function parameter "resize" to "browserResized" in
adjustBrowseHeight and updateSliderHandle.
- In browse.js, remove unecessary comments in calculateGalleryHeight.
- In browse.js, fix the updateSliderHandle indentation.

juzen2003 committed:
- In browse.js, in updateSliderHandle function line 707, change "resize"
  to "browserResized".

juzen2003 committed:
- In browse.js, rename setScrollbarOnSlide function to setScrollbarPosition.

juzen2003 committed:
- In browse.js, update event handler function for "request.infiniteScroll"
  in initInfiniteScroll to properly remove spinner (line 1503-1512).

juzen2003 committed:
- In browse.js, update updateBrowseNav to update a global fading flag, it
  is used to determine if the next updateBrowseNav should be called in event
  handler function when clicking ".op-browse-view" (line 101).
- In browse.js, update getBrowseView to take in a "tab" parameter so that
  we know which tab we are at when calling this function.
- In browse.js, update getScrollContainerClass to take in a "tab" parameter
  so that we know which tab we are at when calling this function.
- In browse.js, in loadData, pass in view when calling getStartObsLabel and
  getScrollContainerClass.
- In browse.js, in renderGalleryAndTable, pass in view when calling
  getScrollContainerClass.

juzen2003 committed:
- In browse.js, in updateSliderHandle, update the slider max value to be
  the last page's startObs. It will be used as the max value for slider.
